File: zig/src/slab.zig
//! The bytes a field points into, and the one route every comparison reads them
//! through.
//!
//! A field is an offset and a length, so something has to say what those address
//! and how the bytes are escaped. For CSV and JSON that is the mapped file; for
//! Parquet it is an arena of decoded values, which is the same thing to everyone
//! downstream. The escape rule rides along because it follows the format: CSV
//! doubles a quote, JSON puts a backslash in front of one, and a decoded Parquet
//! value is already literal.
//!
//! Hashing, equality and decoding all read a field through `logical`, so they
//! cannot come to different conclusions about the same value — the property whose
//! absence produced two silently wrong answers in the Java port.

const std = @import("std");
const f = @import("field.zig");
const Field = f.Field;

pub const Error = error{CannotReadFile};

/// Which escape rule the bytes behind a field follow.
pub const Dialect = enum {
    /// A doubled quote is one quote.
    csv,
    /// A backslash escape, `\uXXXX` included.
    json,
    /// The bytes are the value: nothing to undo. Parquet decodes to this.
    raw,
};

pub const Slab = struct {
    data: []const u8,
    dialect: Dialect = .csv,
    /// Set when the bytes are a mapping, so they can be given back.
    mapping: ?[]align(std.heap.page_size_min) const u8 = null,
    /// Set when the bytes are an arena this process built.
    arena: ?[]u8 = null,
    gpa: ?std.mem.Allocator = null,

    /// Maps a file read-only. The whole file is read once, front to back.
    pub fn map(io: std.Io, path: []const u8) !Slab {
        const file = std.Io.Dir.cwd().openFile(io, path, .{}) catch return Error.CannotReadFile;
        defer file.close(io);
        const size = (file.stat(io) catch return Error.CannotReadFile).size;
        if (size == 0) return Slab{ .data = &[_]u8{} };
        const mapped = std.posix.mmap(
            null,
            size,
            .{ .READ = true },
            .{ .TYPE = .PRIVATE },
            file.handle,
            0,
        ) catch return Error.CannotReadFile;
        return Slab{ .data = mapped, .mapping = mapped };
    }

    /// A slab over bytes this process built: the Parquet reader's arena.
    pub fn owned(gpa: std.mem.Allocator, bytes: []u8, dialect: Dialect) Slab {
        return .{ .data = bytes, .dialect = dialect, .arena = bytes, .gpa = gpa };
    }

    pub fn close(self: *Slab) void {
        if (self.mapping) |m| std.posix.munmap(m);
        if (self.arena) |a| {
            if (self.gpa) |gpa| gpa.free(a);
        }
        self.* = .{ .data = &[_]u8{} };
    }

    /// The field's raw span, still holding whatever escapes it was written with.
    pub fn raw(self: Slab, field: Field) []const u8 {
        if (!f.isReal(field)) return &[_]u8{};
        return self.data[f.offsetOf(field)..][0..f.lenOf(field)];
    }

    /// The field's logical bytes: the value, with its escapes undone.
    pub fn logical(self: Slab, field: Field) Logical {
        const escaped = f.isReal(field) and f.isEscaped(field);
        return .{
            .raw = self.raw(field),
            .dialect = if (escaped) self.dialect else .raw,
        };
    }
};

/// A field's bytes with its escapes undone, one byte at a time and without
/// allocating. A `\uXXXX` escape decodes to as many as four bytes, which is what
/// the small pending buffer is for.
pub const Logical = struct {
    raw: []const u8,
    dialect: Dialect,
    at: usize = 0,
    pending: [4]u8 = undefined,
    pending_len: u8 = 0,
    pending_at: u8 = 0,

    /// True when the bytes are the value, so a caller may take the slice whole.
    pub fn isPlain(self: Logical) bool {
        return self.dialect == .raw;
    }

    pub fn next(self: *Logical) ?u8 {
        if (self.pending_at < self.pending_len) {
            const b = self.pending[self.pending_at];
            self.pending_at += 1;
            return b;
        }
        if (self.at >= self.raw.len) return null;
        const b = self.raw[self.at];
        self.at += 1;
        switch (self.dialect) {
            .raw => return b,
            .csv => {
                // A quote inside a quoted body can only be half of a doubled
                // pair, so the second one is dropped.
                if (b == '"' and self.at < self.raw.len and self.raw[self.at] == '"') self.at += 1;
                return b;
            },
            .json => {
                if (b != '\\' or self.at >= self.raw.len) return b;
                const e = self.raw[self.at];
                self.at += 1;
                return switch (e) {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    'b' => 0x08,
                    'f' => 0x0c,
                    '"' => '"',
                    '\\' => '\\',
                    '/' => '/',
                    'u' => blk: {
                        if (self.takeEscape()) |cp| break :blk self.hold(cp);
                        // Not four hex digits: it was never an escape, so both
                        // bytes are the value.
                        break :blk self.holdByte('\\', 'u');
                    },
                    // Not an escape this dialect defines: both bytes are content.
                    else => self.holdByte('\\', e),
                };
            },
        }
    }

    fn holdByte(self: *Logical, first: u8, second: u8) u8 {
        self.pending[0] = second;
        self.pending_len = 1;
        self.pending_at = 0;
        return first;
    }

    /// Encodes one code point, returning its first byte and holding the rest.
    ///
    /// Written out rather than taken from `std.unicode`, which refuses an
    /// unpaired surrogate: a file carrying one is malformed, but the C, C++ and
    /// Rust ports encode it as written rather than substituting a replacement
    /// character, and two ports that disagree about the bytes of a value would
    /// disagree about its hash.
    fn hold(self: *Logical, cp: u21) u8 {
        var buf: [4]u8 = undefined;
        var n: u8 = 0;
        if (cp < 0x80) {
            buf[0] = @intCast(cp);
            n = 1;
        } else if (cp < 0x800) {
            buf[0] = @intCast(0xC0 | (cp >> 6));
            buf[1] = @intCast(0x80 | (cp & 0x3F));
            n = 2;
        } else if (cp < 0x10000) {
            buf[0] = @intCast(0xE0 | (cp >> 12));
            buf[1] = @intCast(0x80 | ((cp >> 6) & 0x3F));
            buf[2] = @intCast(0x80 | (cp & 0x3F));
            n = 3;
        } else {
            buf[0] = @intCast(0xF0 | (cp >> 18));
            buf[1] = @intCast(0x80 | ((cp >> 12) & 0x3F));
            buf[2] = @intCast(0x80 | ((cp >> 6) & 0x3F));
            buf[3] = @intCast(0x80 | (cp & 0x3F));
            n = 4;
        }
        self.pending = buf;
        self.pending_len = n;
        self.pending_at = 1;
        return buf[0];
    }

    /// The code point of the `\u` escape whose digits start at `at`, with a
    /// following low surrogate folded in. Leaves `at` past what it consumed.
    fn takeEscape(self: *Logical) ?u21 {
        const hi = self.takeHex4() orelse return null;
        if (hi < 0xD800 or hi > 0xDBFF) return @intCast(hi);
        // A surrogate pair is one code point written as two escapes.
        if (self.at + 1 < self.raw.len and self.raw[self.at] == '\\' and self.raw[self.at + 1] == 'u') {
            const save = self.at;
            self.at += 2;
            if (self.takeHex4()) |lo| {
                if (lo >= 0xDC00 and lo <= 0xDFFF) {
                    return @intCast(0x10000 + ((hi - 0xD800) << 10) + (lo - 0xDC00));
                }
            }
            self.at = save;
        }
        return @intCast(hi);
    }

    fn takeHex4(self: *Logical) ?u32 {
        if (self.at + 4 > self.raw.len) return null;
        var cp: u32 = 0;
        for (self.raw[self.at..][0..4]) |d| {
            const v: u32 = switch (d) {
                '0'...'9' => d - '0',
                'a'...'f' => d - 'a' + 10,
                'A'...'F' => d - 'A' + 10,
                else => return null,
            };
            cp = cp * 16 + v;
        }
        self.at += 4;
        return cp;
    }
};

/// Whether two fields hold the same logical bytes, without decoding either.
pub fn sameBytes(a: Slab, x: Field, b: Slab, y: Field) bool {
    // Ask the two field words directly rather than building two iterators to ask
    // them. `logical()` already reports a field with no escape as plain, so the
    // memcmp below was always the path a CSV file took -- but reaching it cost
    // two `Logical` values, each slicing the slab and carrying a four-byte
    // pending buffer, to answer a question two bit tests answer. For a
    // nine-byte field that setup was most of the comparison: 0.76 billion
    // instructions of a 4.3 billion run, down to 0.53 billion.
    if (!f.isEscaped(x) and !f.isEscaped(y)) return std.mem.eql(u8, a.raw(x), b.raw(y));
    var lx = a.logical(x);
    var ly = b.logical(y);
    while (true) {
        const cx = lx.next();
        const cy = ly.next();
        if (cx == null and cy == null) return true;
        if (cx == null or cy == null) return false;
        if (cx.? != cy.?) return false;
    }
}

fn collect(text: []const u8, dialect: Dialect, out: []u8) []const u8 {
    const slab = Slab{ .data = text, .dialect = dialect };
    var it = slab.logical(f.pack(0, text.len, true));
    var n: usize = 0;
    while (it.next()) |b| : (n += 1) out[n] = b;
    return out[0..n];
}

test "csv drops the second quote of a pair" {
    var buf: [64]u8 = undefined;
    try std.testing.expectEqualStrings("a\"b", collect("a\"\"b", .csv, &buf));
}

test "json undoes its escapes" {
    var buf: [64]u8 = undefined;
    try std.testing.expectEqualStrings("a\nb\"c\\d", collect("a\\nb\\\"c\\\\d", .json, &buf));
}

test "a unicode escape and the character itself are the same value" {
    var buf: [64]u8 = undefined;
    try std.testing.expectEqualStrings("caf\u{e9}", collect("caf\\u00e9", .json, &buf));
    try std.testing.expectEqualStrings("\u{1f600}", collect("\\ud83d\\ude00", .json, &buf));
    // Not four hex digits, so it was never an escape.
    try std.testing.expectEqualStrings("\\uZZ", collect("\\uZZ", .json, &buf));
    // A lone high surrogate is encoded as written, which is what the other ports
    // do, and must not eat the character after it.
    try std.testing.expectEqualStrings("\xed\xa0\xbdx", collect("\\ud83dx", .json, &buf));
}

File: rust/src/engine/turbo/slab.rs
//! The bytes a field points into, and the one route every comparison reads them
//! through.
//!
//! A field is an offset and a length, so something has to say what those address
//! and how the bytes are escaped. For CSV and JSON that is the mapped file; for
//! Parquet it is an arena of decoded values, which is the same thing to everyone
//! downstream. The escape rule rides along because it follows the format: CSV
//! doubles a quote, JSON puts a backslash in front of one, and a decoded Parquet
//! value is already literal.
//!
//! Hashing, equality and decoding all read a field through [`Slab::logical`], so
//! they cannot come to different conclusions about the same value — the property
//! whose absence produced two silently wrong answers in the Java port.

use std::fs::File;
use std::path::Path;

use memmap2::Mmap;

use super::field::{Field, is_escaped, is_real, len_of, offset_of};
use crate::error::{Error, Result};

/// Which escape rule the bytes behind a field follow.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum Dialect {
    /// A doubled quote is one quote.
    Csv,
    /// A backslash escape, `\uXXXX` included.
    Json,
    /// The bytes are the value: nothing to undo. Parquet decodes to this.
    Raw,
}

/// Where the bytes came from. A mapping is paged in on demand and evictable; an
/// arena is what a columnar format leaves behind once its pages are decoded.
enum Bytes {
    Mapped { _file: File, map: Mmap },
    Owned(Vec<u8>),
}

pub(super) struct Slab {
    bytes: Bytes,
    dialect: Dialect,
}

impl Slab {
    /// Maps a file read-only, with a hint that it will be read front to back.
    pub(super) fn map(path: &Path) -> Result<Self> {
        let file = File::open(path)
            .map_err(|e| Error::new(format!("cannot read {}: {e}", path.display())))?;
        // Safety: the file is opened read-only and not modified while mapped.
        // A concurrent truncation would be a torn read, which is the documented
        // hazard of every mapping engine here and of DuckDB's reader too.
        let map = unsafe { Mmap::map(&file) }
            .map_err(|e| Error::new(format!("cannot map {}: {e}", path.display())))?;
        let _ = map.advise(memmap2::Advice::Sequential);
        Ok(Slab {
            bytes: Bytes::Mapped { _file: file, map },
            dialect: Dialect::Csv,
        })
    }

    /// A slab over bytes this process built: the Parquet reader's arena.
    pub(super) fn owned(data: Vec<u8>, dialect: Dialect) -> Self {
        Slab {
            bytes: Bytes::Owned(data),
            dialect,
        }
    }

    pub(super) fn set_dialect(&mut self, dialect: Dialect) {
        self.dialect = dialect;
    }

    pub(super) fn dialect(&self) -> Dialect {
        self.dialect
    }

    pub(super) fn data(&self) -> &[u8] {
        match &self.bytes {
            Bytes::Mapped { map, .. } => map,
            Bytes::Owned(v) => v,
        }
    }

    /// The field's raw span, still holding whatever escapes it was written with.
    pub(super) fn raw(&self, f: Field) -> &[u8] {
        if !is_real(f) {
            return &[];
        }
        let (at, len) = (offset_of(f), len_of(f));
        &self.data()[at..at + len]
    }

    /// The field's logical bytes: the value, with its escapes undone.
    pub(super) fn logical(&self, f: Field) -> LogicalBytes<'_> {
        let escaped = is_real(f) && is_escaped(f);
        LogicalBytes {
            raw: self.raw(f),
            dialect: if escaped { self.dialect } else { Dialect::Raw },
            at: 0,
            pending: [0; 4],
            pending_len: 0,
            pending_at: 0,
        }
    }
}

/// A field's bytes with its escapes undone, one byte at a time and without
/// allocating. A `\uXXXX` escape decodes to as many as four bytes, which is what
/// the small pending buffer is for.
pub(super) struct LogicalBytes<'a> {
    raw: &'a [u8],
    dialect: Dialect,
    at: usize,
    pending: [u8; 4],
    pending_len: u8,
    pending_at: u8,
}

impl LogicalBytes<'_> {
    /// True when the bytes are the value, so a caller may take the slice whole.
    pub(super) fn is_plain(&self) -> bool {
        self.dialect == Dialect::Raw
    }

    /// Encodes one code point, returning its first byte and holding the rest.
    ///
    /// The encoding is done by hand rather than through `char`, which refuses an
    /// unpaired surrogate: a file carrying one is malformed, but the C, C++ and
    /// Zig ports encode it rather than substituting a replacement character, and
    /// two ports that disagree about the bytes of a value would disagree about
    /// its hash.
    fn hold(&mut self, cp: u32) -> u8 {
        let mut buf = [0u8; 4];
        let n = match cp {
            0..=0x7F => {
                buf[0] = cp as u8;
                1
            }
            0x80..=0x7FF => {
                buf[0] = 0xC0 | (cp >> 6) as u8;
                buf[1] = 0x80 | (cp & 0x3F) as u8;
                2
            }
            0x800..=0xFFFF => {
                buf[0] = 0xE0 | (cp >> 12) as u8;
                buf[1] = 0x80 | ((cp >> 6) & 0x3F) as u8;
                buf[2] = 0x80 | (cp & 0x3F) as u8;
                3
            }
            _ => {
                buf[0] = 0xF0 | (cp >> 18) as u8;
                buf[1] = 0x80 | ((cp >> 12) & 0x3F) as u8;
                buf[2] = 0x80 | ((cp >> 6) & 0x3F) as u8;
                buf[3] = 0x80 | (cp & 0x3F) as u8;
                4
            }
        };
        self.pending = buf;
        self.pending_len = n;
        self.pending_at = 1;
        buf[0]
    }
}

impl Iterator for LogicalBytes<'_> {
    type Item = u8;

    fn next(&mut self) -> Option<u8> {
        if self.pending_at < self.pending_len {
            let b = self.pending[self.pending_at as usize];
            self.pending_at += 1;
            return Some(b);
        }
        let b = *self.raw.get(self.at)?;
        self.at += 1;
        match self.dialect {
            Dialect::Raw => Some(b),
            Dialect::Csv => {
                // A quote inside a quoted body can only be half of a doubled
                // pair, so the second one is dropped.
                if b == b'"' && self.raw.get(self.at) == Some(&b'"') {
                    self.at += 1;
                }
                Some(b)
            }
            Dialect::Json => {
                if b != b'\\' {
                    return Some(b);
                }
                let Some(&e) = self.raw.get(self.at) else {
                    return Some(b);
                };
                self.at += 1;
                Some(match e {
                    b'n' => b'\n',
                    b't' => b'\t',
                    b'r' => b'\r',
                    b'b' => 0x08,
                    b'f' => 0x0c,
                    b'"' => b'"',
                    b'\\' => b'\\',
                    b'/' => b'/',
                    b'u' => match self.take_escape() {
                        // A code point becomes UTF-8 rather than staying as the
                        // six bytes it was written as: a JSON writer may spell a
                        // character either way, and both spellings have to
                        // compare equal.
                        Some(cp) => self.hold(cp),
                        None => {
                            // Not four hex digits: it was never an escape, so
                            // both bytes are the value.
                            self.pending = [b'u', 0, 0, 0];
                            self.pending_len = 1;
                            self.pending_at = 0;
                            b'\\'
                        }
                    },
                    // Not an escape this dialect defines: both bytes are content.
                    other => {
                        self.pending = [other, 0, 0, 0];
                        self.pending_len = 1;
                        self.pending_at = 0;
                        b'\\'
                    }
                })
            }
        }
    }
}

impl LogicalBytes<'_> {
    /// The code point of the `\u` escape whose digits start at `self.at`, with a
    /// following low surrogate folded in. Leaves `at` past what it consumed.
    fn take_escape(&mut self) -> Option<u32> {
        let hi = self.take_hex4()?;
        if !(0xD800..=0xDBFF).contains(&hi) {
            return Some(hi);
        }
        // A surrogate pair is one code point written as two escapes.
        if self.raw.get(self.at) == Some(&b'\\') && self.raw.get(self.at + 1) == Some(&b'u') {
            let save = self.at;
            self.at += 2;
            match self.take_hex4() {
                Some(lo) if (0xDC00..=0xDFFF).contains(&lo) => {
                    return Some(0x10000 + ((hi - 0xD800) << 10) + (lo - 0xDC00));
                }
                _ => self.at = save,
            }
        }
        Some(hi)
    }

    fn take_hex4(&mut self) -> Option<u32> {
        let digits = self.raw.get(self.at..self.at + 4)?;
        let mut cp = 0u32;
        for &d in digits {
            cp = cp * 16 + (d as char).to_digit(16)?;
        }
        self.at += 4;
        Some(cp)
    }
}

/// Whether two fields hold the same logical bytes, without decoding either.
pub(super) fn same_bytes(a: &Slab, x: Field, b: &Slab, y: Field) -> bool {
    // Ask the two field words directly rather than building two iterators to ask
    // them. `logical()` already reports a field with no escape as plain, so the
    // memcmp below was always the path a CSV file took -- but reaching it cost
    // two `LogicalBytes` values, each slicing the slab and carrying a four-byte
    // pending buffer, to answer a question two bit tests answer. For a
    // nine-byte field that setup was most of the comparison.
    if !is_escaped(x) && !is_escaped(y) {
        return a.raw(x) == b.raw(y);
    }
    a.logical(x).eq(b.logical(y))
}

/// The field decoded. Only the report sections and the normalising paths call
/// this; the fast path never builds a string for a cell at all.
pub(super) fn text_of(slab: &Slab, f: Field) -> String {
    let logical = slab.logical(f);
    if logical.is_plain() {
        return String::from_utf8_lossy(slab.raw(f)).into_owned();
    }
    let bytes: Vec<u8> = logical.collect();
    String::from_utf8_lossy(&bytes).into_owned()
}

#[cfg(test)]
mod tests {
    use super::super::field::pack;
    use super::*;

    fn bytes(text: &[u8], dialect: Dialect) -> Vec<u8> {
        let slab = Slab::owned(text.to_vec(), dialect);
        slab.logical(pack(0, text.len() as u64, true)).collect()
    }

    fn decode(text: &[u8], dialect: Dialect) -> String {
        String::from_utf8(bytes(text, dialect)).expect("valid utf-8")
    }

    #[test]
    fn csv_drops_the_second_quote_of_a_pair() {
        assert_eq!(decode(b"a\"\"b", Dialect::Csv), "a\"b");
    }

    #[test]
    fn json_undoes_its_escapes() {
        assert_eq!(decode(br#"a\nb\"c\\d"#, Dialect::Json), "a\nb\"c\\d");
    }

    #[test]
    fn a_unicode_escape_and_the_character_itself_are_the_same_value() {
        // A writer may spell a character either way, so both spellings have to
        // decode to the same bytes.
        assert_eq!(decode(br"caf\u00e9", Dialect::Json), "caf\u{e9}");
        assert_eq!(decode(br"\ud83d\ude00", Dialect::Json), "\u{1f600}");
        // Not four hex digits, so it was never an escape.
        assert_eq!(decode(br"\uZZ", Dialect::Json), r"\uZZ");
        // A lone high surrogate has no pair to fold in and must not eat what
        // follows it. It is encoded as written, which is what the other ports
        // do; substituting a replacement character here would make the same
        // file hash differently in two ports.
        assert_eq!(bytes(br"\ud83dx", Dialect::Json), b"\xed\xa0\xbdx");
    }
}
